File: Problems/1254. (M) Number of Closed Islands/Python3/solution.py
class Solution:
  def closedIsland(self, grid: List[List[int]]) -> int:
    # iterative DFS

    m = len(grid)
    n = len(grid[0])
    # shared across traversals
    visited = set()

    def dfs(y, x):
      # sanity check
      if (y, x) in visited:
        return False
      border = 0
      nodes = []
      nodes.append((y, x))
      while nodes:
        cur = nodes.pop()
        if cur in visited:
          continue
        visited.add(cur)
        c_y, c_x = cur
        # raise flag if land square on border
        if not c_y or c_y == m-1 or not c_x or c_x == n-1:
          border = 1
        nxt = [(c_y+1,c_x), (c_y,c_x+1), (c_y-1,c_x), (c_y,c_x-1)]
        for n_y, n_x in nxt:
          if 0 <= n_y < m and 0 <= n_x < n and not grid[n_y][n_x]:
            nodes.append((n_y,n_x))

      return True if not border else False
    
    ret = 0
    for i in range(m):
      for j in range(n):
        if not grid[i][j] and dfs(i, j):
          ret += 1
    
    return ret

File: Problems/1254. (M) Number of Closed Islands/Java/solution.java
class Solution {
  // iterative DFS
  public int closedIsland(int[][] grid) {
    Set<Pair<Integer,Integer>> visited = new HashSet();
    int ret = 0;
    for (int i = 0; i < grid.length; i++) {
      for (int j = 0; j < grid[0].length; j++) {
        if (grid[i][j] == 0 && dfs(i, j, grid, visited))
          ret++;
      }
    }
    return ret;
  }
  private boolean dfs(int y, int x, int[][] grid, Set<Pair<Integer,Integer>> visited) {
    // sanity check
    if (visited.contains(new Pair<Integer,Integer>(y,x)))
      return false;
    boolean border = false;
    Deque<Pair<Integer,Integer>> nodes = new ArrayDeque();
    nodes.push(new Pair<Integer,Integer>(y,x));
    while (nodes.size() > 0) {
      Pair<Integer,Integer> cur = nodes.pop();
      if (visited.contains(cur))
        continue;
      visited.add(cur);
      int c_y = cur.getKey();
      int c_x = cur.getValue();
      // raise flag if land square is in border
      if (c_y == 0 || c_y == grid.length-1 || c_x == 0 || c_x == grid[0].length-1)
        border = true;
      int[][] next = {{c_y+1,c_x}, {c_y,c_x+1}, {c_y-1,c_x}, {c_y,c_x-1}};
      for (int[] coord: next) {
        int n_y = coord[0], n_x = coord[1];
        if (0 <= n_y && n_y < grid.length && 0 <= n_x && n_x < grid[0].length &&
          grid[n_y][n_x] == 0)
          nodes.push(new Pair<Integer,Integer>(coord[0],coord[1]));
      }
    }
    return !border;
  }
}
